### PR TITLE
feat: update position utility and improve resize detection

### PR DESCRIPTION
**Pull request by cal-smith**
- updates `@carbon/utils-position` to `1.1.2`
- adds `AnimationFrameService` and `AnimationFrameServiceSingleton` so we have a consistent and effeciant way of hooking into `requestAnimationFrame`
- add `ElementService` to simplify visibility checks (and more in future)
- refactor `dropdown`, `dropdown.service`, and `dialog` to use `ElementService` and `AnimationFrameService`

### CODE REVIEW

File: src/utils/scroll.ts [hunk 1]
```diff
@@ -30,6 +30,17 @@ export const isVisibleInContainer = (element: HTMLElement, container: HTMLElemen
 	return elementRect.bottom <= containerRect.bottom && elementRect.top >= containerRect.top;
 };
 
+export const getScrollableParents = (node: HTMLElement) => {
+	const elements = [document.body];
```
Comment (zvonimirfras) [inline src/utils/scroll.ts]: Check if we're in the browser to use `document`?

File: src/dialog/overflow-menu/overflow-menu-pane.component.ts [hunk 2]
```diff
@@ -30,8 +31,12 @@ import { ExperimentalService } from "./../../experimental.module";
 	`
 })
 export class OverflowMenuPane extends Dialog implements AfterViewInit {
-	constructor(protected elementRef: ElementRef, protected i18n: I18n, protected experimental: ExperimentalService) {
-		super(elementRef);
+	constructor(
+		protected elementRef: ElementRef,
+		protected i18n: I18n,
+		protected experimental: ExperimentalService,
+		protected elementService: ElementService) {
```
[the remark this answers is not in the document]
Comment (cal-smith) [reply src/dialog/overflow-menu/overflow-menu-pane.component.ts]: Sure 👍

File: src/utils/scroll.ts (hunk 1 above)
Comment (cal-smith) [reply src/utils/scroll.ts]: Shouldn't be necessary given that we only query when the components have live DOM ... Really we need a proper strategy to make the components SSR safe, as-is there's a few that won't even initialize

File: src/utils/element.service.ts [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+	}
+
+	visibility(target: HTMLElement, parentElement: HTMLElement = target): Observable<{ visible: boolean, frame: number }> {
+		const scrollableParents = getScrollableParents(parentElement);
+		return this.tick.pipe(map(frame => {
+			for (const parent of scrollableParents) {
+				if (!isVisibleInContainer(target, parent)) {
+					return {
+						visible: false,
+						frame
```
Comment (Donisius) [inline src/utils/element.service.ts]: I don't see a reason why we need to return the frame along with visibility.

File: src/dialog/dialog.component.ts [hunk 4]
```diff
@@ -63,13 +64,15 @@ export class Dialog implements OnInit, AfterViewInit, OnDestroy {
 	/**
 	 * `Subscription` used to update placement in the event of a window resize.
 	 */
-	protected resizeSubscription: Subscription;
+	protected resizeSubscription: Subscription = new Subscription();
```
Comment (theiliad) [inline src/dialog/dialog.component.ts]: you can remove `: Subscription` now that the type is implicitly defined through `new Subscription()`

File: src/dialog/dialog.component.ts [hunk 5]
```diff
@@ -63,13 +64,15 @@ export class Dialog implements OnInit, AfterViewInit, OnDestroy {
...
 	/**
 	 * Subscription to all the scrollable parents `scroll` event
 	 */
 	// add a new subscription temporarily so that contexts (such as tests)
 	// that don't run ngAfterViewInit have something to unsubscribe in ngOnDestroy
 	protected scrollSubscription: Subscription = new Subscription();
```
Comment (theiliad) [inline src/dialog/dialog.component.ts]: remove `: Subscription`
```suggestion
	protected scrollSubscription = new Subscription();
```

File: src/dialog/dialog.component.ts (hunk 4 above)
Comment (theiliad) [inline src/dialog/dialog.component.ts]:
```suggestion
	protected resizeSubscription = new Subscription();
```

File: src/dialog/dialog.component.ts [hunk 6]
```diff
@@ -101,9 +108,12 @@ export class Dialog implements OnInit, AfterViewInit, OnDestroy {
...
 		this.data = this.dialogConfig.data;
 
-		this.resizeSubscription = Dialog.resizeObservable.subscribe(() => {
-			this.placeDialog();
-		});
+		// fallback if elementService isn't available
+		if (!this.elementService) {
+			this.resizeSubscription = Dialog.resizeObservable.subscribe(() => {
+				this.placeDialog();
+			});
```
Comment (theiliad) [inline src/dialog/dialog.component.ts]: Why not do:
```js
this.resizeSubscription = Dialog.resizeObservable.subscribe(this.placeDialog);
```

File: src/dropdown/dropdown.service.ts [hunk 7]
```diff
@@ -21,14 +21,17 @@ export class DropdownService {
 	protected menuInstance: HTMLElement;
 
 	/**
-	 * Maintains an Event Observable Subscription for tracking window resizes.
-	 * Window resizing is tracked if the `Dropdown` is appended to the body, otherwise it does not need to be supported.
+	 * Maintains an Event Observable Subscription for the global requestAnimationFrame.
+	 * requestAnimationFrame is tracked only if the `Dropdown` is appended to the body otherwise we don't need it
 	 */
-	protected resize: Subscription;
+	protected animationFrameSubscription: Subscription = new Subscription();
```
Comment (theiliad) [inline src/dropdown/dropdown.service.ts]:
```suggestion
	protected animationFrameSubscription = new Subscription();
```

File: src/dropdown/dropdown.service.ts [hunk 8]
```diff
@@ -56,10 +59,9 @@ export class DropdownService {
 
 		this.menuInstance = dropdownWrapper;
 
-		this.positionDropdown(parentRef, dropdownWrapper);
-		this.resize = fromEvent(window, "resize")
-			.pipe(throttleTime(100))
-			.subscribe(() => this.positionDropdown(parentRef, dropdownWrapper));
+		this.animationFrameSubscription = this.animationFrameService.tick.subscribe(() => {
+			this.positionDropdown(parentRef, dropdownWrapper);
+		});
```
Comment (theiliad) [inline src/dropdown/dropdown.service.ts]: What about?
```js
this.animationFrameSubscription = this.animationFrameService.tick
   .subscribe(() => this.positionDropdown(parentRef, dropdownWrapper));
```

File: src/dialog/dialog.component.ts (hunk 6 above)
Comment (cal-smith) [reply src/dialog/dialog.component.ts]: I don't _think_ we'll run into scoping issues with that ... I'll give it a try 👍

File: src/dropdown/dropdown.service.ts (hunk 8 above)
Comment (cal-smith) [reply src/dropdown/dropdown.service.ts]: More or less the same thing ... Currently it's marginally more readable to my eyes

**Pull request merged**